### PR TITLE
ARROW-6486: [Python] Allow subclassing & monkey-patching of Table

### PR DESCRIPTION
**Pull request by ARF1**
Currently, many classes in ``pyarrow`` behave strangely to the Python user: they are neither subclassable not monkey-patchable.

```
>>> import pyarrow as pa
>>> class MyTable(pa.Table):
... pass
...
>>> table = MyTable.from_arrays([], [])
>>> type(table)
<class 'pyarrow.lib.Table'>
```

The factory method did not return an instance of our subclass...

Never mind, let's monkey-patch Table:
```
>>> pa.TableOriginal = pa.Table
>>> pa.Table = MyTable
>>> table = pa.Table.from_arrays([], [])
>>> type(table)
<class 'pyarrow.lib.Table'>
```

OK, that did not work either.

Let's be sneaky:
```

>>> table._class_ = MyTable
Traceback (most recent call last):
{{ File "<stdin>", line 1, in <module>}}
TypeError: _class_ assignment only supported for heap types or ModuleType subclasses
>>>
```

 

There is currently no way to modify or extend the behaviour of a Table instance. Users can use only what pyarrow provides out of the box. - This is likely to be a source of frustration for many python users.

 

The attached PR remedies this for the Table class:

```
>>> import pyarrow as pa
>>> class MyTable(pa.Table):
... pass
...
>>> table = MyTable.from_arrays([], [])
>>> type(table)
<class '_main_.MyTable'>
>>>
>>> pa.TableOriginal = pa.Table
>>> pa.Table = MyTable
>>> table = pa.Table.from_arrays([], [])
>>> type(table)
<class '_main_.MyTable'>
>>>
```

 

Ideally, these modifications would be extended to the other cython-defined classes of pyarrow, but given that Table is likely to be the interface that most users begin their interaction with, I thought this would be a good start.

Keeping the changes limited to a single class should also keep merge conflicts manageable.

### CODE REVIEW

**Comment by wesm** [comment]
I am not very excited about this. These classes are intended to be developer-facing, not user-facing. If you desire a different user interface, you should use composition rather than inheritance.

**Comment by jorisvandenbossche** [comment]
I agree with Wes. As I mentioned on the issue: from experience with pandas, properly enabling subclassing of a complex data structure adds quite some complexity to the code with lots of corner cases. So I am not sure this is something we want to do for pyarrow Table. As Wes said, there are alternatives to subclassing if you want to use pyarrow.Table's functionality in a custom interface.

**Comment by wesm** [comment]
FWIW, the same is also true of NumPy. Subclasses of ndarray have never really worked out the way they were intended, and I think it would have been better to not offer the feature in the first place.

File: python/pyarrow/lib.pxd [hunk 1]
```diff
@@ -490,7 +490,8 @@ cdef public object pyarrow_wrap_field(const shared_ptr[CField]& field)
 cdef public object pyarrow_wrap_resizable_buffer(
     const shared_ptr[CResizableBuffer]& buf)
 cdef public object pyarrow_wrap_schema(const shared_ptr[CSchema]& type)
-cdef public object pyarrow_wrap_table(const shared_ptr[CTable]& ctable)
+cdef public object pyarrow_wrap_table(
+    object table_class, const shared_ptr[CTable]& ctable)
```
Comment (pitrou) [inline python/pyarrow/lib.pxd]: I wonder if it's possible to keep the old overload as well (the one-argument form). Does Cython allow this?

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -1481,6 +1481,8 @@ cdef class Table(_PandasConvertible):
 
         return table
 
+Table = _Table
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: I don't understand why you're doing this. The `__name__` of the class will be different, pickle compatibility may be broken... I'd rather keep the original name (i.e. `Table`).

File: python/pyarrow/tests/test_table.py [hunk 3]
```diff
@@ -991,3 +991,60 @@ def test_table_function_unicode_schema():
...
+    schema = pa.schema([
+        pa.field('a', pa.int64()),
+        pa.field('b', pa.float64()),
+    ])
+    batch = pa.record_batch([pa.array([1]), pa.array([3.14])],
+                            names=['a', 'b'])
+    table = MyTable.from_batches([batch], schema)
+    assert type(table) == MyTable
+
+    import pandas as pd
```
Comment (pitrou) [inline python/pyarrow/tests/test_table.py]: Use `@pytest.mark.pandas` if a test requires Pandas, as it should be possible to run the test suite without Pandas installed.

**Review by pitrou — COMMENTED** [review]
Thanks for doing this. I wonder if we should break the existing API like this. See comment.

**Comment by wesm** [comment]
I'm revising my position on this to -1.

**Comment by wesm** [comment]
I'm closing for maintainability concerns. We can discuss more on dev@

**Pull request closed without merging**